server: TestServeIndexHTML fails under Go 1.17.6

First observed in https://github.com/cockroachdb/cockroach/pull/74655:

```
=== RUN   TestServeIndexHTML
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/5af90cb183adf3be59aa2645ef0012f2/logTestServeIndexHTML2606840330
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestServeIndexHTML
    server_test.go:945: -- test log scope end --
    server_test.go:945: Leaked goroutine: goroutine 67965 [select]:
        net/http.setRequestCancel.func4()
          GOROOT/src/net/http/client.go:398 +0x94
        created by net/http.setRequestCancel
          GOROOT/src/net/http/client.go:397 +0x43e
--- FAIL: TestServeIndexHTML (6.50s)
```

This blocks the Go upgrade (#74617).

--- a/pkg/server/server_test.go
+++ b/pkg/server/server_test.go
@@ -806,6 +806,8 @@ func TestServeIndexHTML(t *testing.T) {
 </html>
 `
 
+	ctx := context.Background()
+
 	linkInFakeUI := func() {
 		ui.HaveUI = true
 	}
@@ -821,26 +823,20 @@ func TestServeIndexHTML(t *testing.T) {
 			// In test servers, web sessions are required by default.
 			DisableWebSessionAuthentication: true,
 		})
-		defer s.Stopper().Stop(context.Background())
+		defer s.Stopper().Stop(ctx)
 		tsrv := s.(*TestServer)
 
 		client, err := tsrv.GetHTTPClient()
-		if err != nil {
-			t.Fatal(err)
-		}
+		require.NoError(t, err)
 
 		t.Run("short build", func(t *testing.T) {
 			resp, err := client.Get(s.AdminURL())
-			if err != nil {
-				t.Fatal(err)
-			}
-			if resp.StatusCode != 200 {
-				t.Fatalf("expected status code 200; got %d", resp.StatusCode)
-			}
+			require.NoError(t, err)
+			require.Equal(t, 200, resp.StatusCode)
+
 			respBytes, err := ioutil.ReadAll(resp.Body)
-			if err != nil {
-				t.Fatal(err)
-			}
+			require.NoError(t, err)
+
 			respString := string(respBytes)
 			expected := fmt.Sprintf(`<!DOCTYPE html>
 <title>CockroachDB</title>
@@ -848,25 +844,19 @@ Binary built without web UI.
 <hr>
 <em>%s</em>`,
 				build.GetInfo().Short())
-			if respString != expected {
-				t.Fatalf("expected %s; got %s", expected, respString)
-			}
+			require.Equal(t, expected, respString)
 		})
 
 		t.Run("non-short build", func(t *testing.T) {
 			linkInFakeUI()
 			defer unlinkFakeUI()
 			resp, err := client.Get(s.AdminURL())
-			if err != nil {
-				t.Fatal(err)
-			}
-			if resp.StatusCode != 200 {
-				t.Fatalf("expected status code 200; got %d", resp.StatusCode)
-			}
+			require.NoError(t, err)
+			require.Equal(t, 200, resp.StatusCode)
+
 			respBytes, err := ioutil.ReadAll(resp.Body)
-			if err != nil {
-				t.Fatal(err)
-			}
+			require.NoError(t, err)
+
 			respString := string(respBytes)
 			expected := fmt.Sprintf(
 				htmlTemplate,
@@ -877,9 +867,7 @@ Binary built without web UI.
 					1,
 				),
 			)
-			if respString != expected {
-				t.Fatalf("expected %s; got %s", expected, respString)
-			}
+			require.Equal(t, expected, respString)
 		})
 	})
 
@@ -887,17 +875,13 @@ Binary built without web UI.
 		linkInFakeUI()
 		defer unlinkFakeUI()
 		s, _, _ := serverutils.StartServer(t, base.TestServerArgs{})
-		defer s.Stopper().Stop(context.Background())
+		defer s.Stopper().Stop(ctx)
 		tsrv := s.(*TestServer)
 
 		loggedInClient, err := tsrv.GetAdminAuthenticatedHTTPClient()
-		if err != nil {
-			t.Fatal(err)
-		}
+		require.NoError(t, err)
 		loggedOutClient, err := tsrv.GetHTTPClient()
-		if err != nil {
-			t.Fatal(err)
-		}
+		require.NoError(t, err)
 
 		cases := []struct {
 			client http.Client
@@ -923,23 +907,28 @@ Binary built without web UI.
 			},
 		}
 
-		for _, testCase := range cases {
-			resp, err := testCase.client.Get(s.AdminURL())
-			if err != nil {
-				t.Fatal(err)
-			}
-			if resp.StatusCode != 200 {
-				t.Fatalf("expected status code 200; got %d", resp.StatusCode)
-			}
-			respBytes, err := ioutil.ReadAll(resp.Body)
-			if err != nil {
-				t.Fatal(err)
-			}
-			respString := string(respBytes)
-			expected := fmt.Sprintf(htmlTemplate, testCase.json)
-			if respString != expected {
-				t.Fatalf("expected %s; got %s", expected, respString)
-			}
+		for i, testCase := range cases {
+			t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+				req, err := http.NewRequestWithContext(ctx, "GET", s.AdminURL(), nil)
+				require.NoError(t, err)
+
+				// Work around this go runtime bug: https://github.com/golang/go/issues/50652
+				cancelCh := make(chan struct{})
+				//lint:ignore SA1019 need this until go bug is resolved
+				req.Cancel = cancelCh
+				defer func() { close(cancelCh) }()
+
+				resp, err := testCase.client.Do(req)
+				require.NoError(t, err)
+				require.Equal(t, 200, resp.StatusCode)
+
+				respBytes, err := ioutil.ReadAll(resp.Body)
+				require.NoError(t, err)
+
+				respString := string(respBytes)
+				expected := fmt.Sprintf(htmlTemplate, testCase.json)
+				require.Equal(t, expected, respString)
+			})
 		}
 	})
 }

